Add config flow for nuki

pree opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Implements config_flow into the nuki integration. 

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Link to documentation pull request: TODO

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->
<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

pree committed:
implemented configflow for nuki

Kane610 commented on homeassistant/components/nuki/binary_sensor.py:
```diff
@@ -0,0 +1,82 @@
+    ATTR_NUKI_ID,
+    DATA_COORDINATOR,
+    DATA_LOCKS,
+    DOMAIN,
```
DOMAIN in platforms are exclusive to the platform itself (binary_sensor)
```suggestion
    DOMAIN as NUKI_DOMAIN,
```

Kane610 commented on homeassistant/components/nuki/__init__.py:
```diff
@@ -1,3 +1,152 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+NUKI_COMPONENTS = ["lock", "binary_sensor"]
```
```suggestion
NUKI_PLATFORMS = ["lock", "binary_sensor"]
```

Kane610 commented on homeassistant/components/nuki/binary_sensor.py:
```diff
@@ -0,0 +1,82 @@
+
+    for lock in hass.data[DOMAIN][entry.entry_id][DATA_LOCKS]:
+        if lock.is_door_sensor_activated:
+            entities.extend([NukiDoorsensorEntity(coordinator, lock)])
```
```suggestion
            entities.append(NukiDoorsensorEntity(coordinator, lock))
```

Kane610 commented on homeassistant/components/nuki/config_flow.py:
```diff
@@ -0,0 +1,78 @@
+        """Handle a flow initiated by the user."""
+        return await self.async_step_init(user_input)
+
+    async def async_step_init(self, user_input):
```
Bette to give it a more appropriate name
```suggestion
    async def async_step_validate(self, user_input):
```

Kane610 commented on homeassistant/components/nuki/lock.py:
```diff
@@ -48,48 +33,46 @@
-    devices = [NukiLockEntity(lock) for lock in bridge.locks]
+    entities = [
+        NukiLockEntity(coordinator, lock)
+        for lock in hass.data[DOMAIN][entry.entry_id][DATA_LOCKS]
```
```suggestion
        for lock in data[DATA_LOCKS]
```

Kane610 commented on homeassistant/components/nuki/binary_sensor.py:
```diff
@@ -0,0 +1,82 @@
+
+    entities = []
+
+    for lock in hass.data[DOMAIN][entry.entry_id][DATA_LOCKS]:
```
```suggestion
    for lock in data[DATA_LOCKS]:
```

Kane610 reviewed (changes requested):
If you introduce a breaking change to the yaml config you can just as well remove it. I suggest you limit this PR to support one configuration and then deprecate the configuration +2 releases down the line and then add multi instance support

bachya commented:
RE: your breaking change, per [ADR-0010](https://github.com/home-assistant/architecture/blob/master/adr/0010-integration-configuration.md), we shouldn't be accepting modifications to configuration-via-YAML functionality. This PR should focus on what's need to add a config entry.

kroimon commented:
Is there a plan on getting this merged?
I might be able to help, but don't fully understand the requested changes. If breaking existing configurations is not an option, how would one proceed with this?

davet2001 commented:
In general the preference is to move towards a GUI based config flow (this is a requirement for new integrations).

Once you have a GUI config flow you can deprecate the YAML.  Functionality changes or new features can be handled that way without breaking changes.

frenck commented:
This PR seems to be running stale. There hasn't been an update from the original author since it initial inception. And it needs to implement ADR-0010 to be accepted.

Closing this PR for now. Please, if you are ready to work on it, do! That is greatly appreciated 👍

pree commented:
Hey @frenck and @pvizeli 

I've just updated this branch and removed the coordinator / doorsensor stuff, so this only implements the config_flow now.
Also I changed it, so it will no longer be a breaking change in the configuration.

Could you please reopen this PR & give it a look? :)

frenck commented:
@pree Please re-open it as a fresh PR 👍 

Your continued work is appreciated!
